src/validation.h

```diff
@@ -189,12 +189,53 @@ void UnlinkPrunedFiles(const std::set<int>& setFilesToPrune);
 /** Prune block files up to a given height */
 void PruneBlockFilesManual(int nManualPruneHeight);
 
-/** (try to) add transaction to memory pool
- * plTxnReplaced will be appended to with all transactions replaced from mempool
- * @param[out] fee_out optional argument to return tx fee to the caller **/
-bool AcceptToMemoryPool(CTxMemPool& pool, TxValidationState &state, const CTransactionRef &tx,
-                        std::list<CTransactionRef>* plTxnReplaced,
-                        bool bypass_limits, bool test_accept=false, CAmount* fee_out=nullptr) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+/**
+* Validation result for a single transaction mempool acceptance.
+* When m_accepted = true, m_replaced_transactions contains a list
+* of replaced transactions and m_base_fees contains the tx fees.
+*/
+struct MempoolAcceptResult {
+    bool m_accepted;
+    TxValidationState m_state;
+
+    // Points to failed tx when m_accepted = false
+    CTransactionRef m_failed_ptx;
+
+    // Valid when m_accepted = true
+    std::optional<std::list<CTransactionRef>> m_replaced_transactions;
+    std::optional<CAmount> m_base_fees;
+
+    /** Constructor for failure case */
+    explicit MempoolAcceptResult(TxValidationState state, CTransactionRef ptx) :
+        m_state(state), m_failed_ptx(ptx) {
+        m_accepted = false;
+        m_replaced_transactions = nullopt;
+        m_base_fees = nullopt;
+    }
+
+    /** Constructor for success case */
+    explicit MempoolAcceptResult(TxValidationState state, std::list<CTransactionRef>&& replaced_txns, CAmount fees) :
+        m_state(state), m_replaced_transactions(std::move(replaced_txns)), m_base_fees(fees) {
+        m_accepted = true;
+        m_failed_ptx = nullptr;
+    }
+};
```

Comment:
Since these ctors are just initializing members, you could do it all in the initializer list:

```suggestion
    /** Constructor for failure case */
    explicit MempoolAcceptResult(TxValidationState state, CTransactionRef ptx) :
        m_accepted(false), m_state(state), m_failed_ptx(ptx), m_replaced_transactions(nullopt),
        m_base_fees(nullopt) {}

    /** Constructor for success case */
    explicit MempoolAcceptResult(TxValidationState state, std::list<CTransactionRef>&& replaced_txns, CAmount fees) :
        m_accepted(true), m_state(state), m_failed_ptx(nullptr),
        m_replaced_transactions(std::move(replaced_txns)), m_base_fees(fees) {}

```